### PR TITLE
Add patterns and grok command

### LINKED ISSUE #814: [FEATURE] Support `grok` command in PPL
**Is your feature request related to a problem?**
PPL currently supports `parse` command that allows user to extract text from a field with regex pattern. But regex is difficult to write while many extracted patterns are likely to be existing patterns. Supporting `grok` can make this easier.

**What solution would you like?**
Allow extracting text using [java-grok](https://github.com/thekrakken/java-grok) similar to current `parse` command.

**What alternatives have you considered?**

**Do you have any additional context?**

### LINKED ISSUE #815: [FEATURE] Support extracting log patterns using PPL
**Is your feature request related to a problem?**
It would be helpful to group similar log lines together, and observe changes happening to different groups of logs, instead of focusing directly on each log line. This can be done by extracting log patterns from each log line. Also see https://github.com/opensearch-project/observability/issues/893.

**What solution would you like?**
Add a command in PPL that lets users extract patterns from a given field.

**What alternatives have you considered?**

**Do you have any additional context?**

### PR DESCRIPTION
**Pull request by joshuali925**
### Description
see #814 , #815 

- add java-grok source code into sql.common
    - [java-grok](https://github.com/thekrakken/java-grok) is an independent library that's used for the PPL's `grok` command, but in the latest release `Grok` is not serializable (and SQL grok expression cannot be pushed down). After discussing with peng I'm tentatively moving the source code into sql.common module.
- add `grok` PPL command, similar to how `parse` currently works with regex
- add `patterns` command, which removes unwanted characters from a log line to create a new field as the pattern of that log line
- `patterns` currently supports default pattern (remove all alphanumerical characters) or user can define what characters should be removed
 
### Issues Resolved
closes #814
closes #815
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
  - [x] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by penghuo** [comment]
@joshuali925 as comments in https://github.com/opensearch-project/sql/issues/814. I think we could add grok as a method in parse command for now.

File: core/src/main/java/org/opensearch/sql/ast/tree/Grok.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+/**
+ * AST node represent Grok operation.
+ */
+@Getter
+@Setter
+@ToString
+@EqualsAndHashCode(callSuper = false)
+@RequiredArgsConstructor
+@AllArgsConstructor
+public class Grok extends UnresolvedPlan {
```
Comment (penghuo) [inline core/src/main/java/org/opensearch/sql/ast/tree/Grok.java:30]: Could we add grok/pattern as the mode in Parse?

File: core/src/main/java/org/opensearch/sql/expression/parse/GrokExpression.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  @Override
+  ExprValue parseValue(ExprValue value) {
+    String rawString = value.stringValue();
+    Match grokMatch = grok.match(rawString);
+    Map<String, Object> capture = grokMatch.capture();
+    Object match = capture.get(identifierStr);
+    if (match != null) {
+      return new ExprStringValue(match.toString());
+    }
+    log.warn("failed to extract pattern {} from input {}", grok.getOriginalGrokPattern(),
```
Comment (penghuo) [inline core/src/main/java/org/opensearch/sql/expression/parse/GrokExpression.java:59]: we should sanitize rawString when logging

Comment (penghuo) [reply core/src/main/java/org/opensearch/sql/expression/parse/GrokExpression.java:59]: is it warn? or debug?

File: core/src/main/java/org/opensearch/sql/expression/ExpressionNodeVisitor.java [hunk 3]
```diff
@@ -11,6 +11,7 @@
 import org.opensearch.sql.expression.conditional.cases.CaseClause;
 import org.opensearch.sql.expression.conditional.cases.WhenClause;
 import org.opensearch.sql.expression.function.FunctionImplementation;
+import org.opensearch.sql.expression.parse.ParseExpression;
```
Comment (penghuo) [inline core/src/main/java/org/opensearch/sql/expression/ExpressionNodeVisitor.java:14]: only import change?

File: core/src/main/java/org/opensearch/sql/expression/parse/GrokExpression.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    return new ExprStringValue("");
+  }
+
+  /**
+   * Get list of derived fields based on parse pattern.
+   *
+   * @param pattern pattern used for parsing
+   * @return list of names of the derived fields
+   */
+  public static List<String> getNamedGroupCandidates(String pattern) {
```
Comment (penghuo) [inline core/src/main/java/org/opensearch/sql/expression/parse/GrokExpression.java:70]: is it only for testing?

File: docs/user/ppl/cmd/patterns.rst [hunk 5]
```diff
@@ -0,0 +1,94 @@
+=============
+parse
```
Comment (penghuo) [inline docs/user/ppl/cmd/patterns.rst:2]: pattern

File: docs/user/ppl/cmd/patterns.rst [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+
+Example 2: Override the existing field
+======================================
+
+The example shows how to override the existing ``address`` field with street number removed using a custom regex filter.
+
+PPL query::
+
+    os> source=accounts | patterns method=regex new_field='address' pattern='\d+ ' address | fields address ;
```
Comment (penghuo) [inline docs/user/ppl/cmd/patterns.rst:65]: the use case is a little bit confuse.

File: core/src/main/java/org/opensearch/sql/ast/tree/Grok.java (hunk 1 above)
Comment (joshuali925) [reply core/src/main/java/org/opensearch/sql/ast/tree/Grok.java]: removed grok and patterns nodes

File: core/src/main/java/org/opensearch/sql/expression/parse/GrokExpression.java (hunk 2 above)
Comment (joshuali925) [reply core/src/main/java/org/opensearch/sql/expression/parse/GrokExpression.java]: changed to debug. rawString is unstructured text not sure how to sanitize, let me know if i should remove it from log

File: core/src/main/java/org/opensearch/sql/expression/ExpressionNodeVisitor.java (hunk 3 above)
Comment (joshuali925) [reply core/src/main/java/org/opensearch/sql/expression/ExpressionNodeVisitor.java:14]: yes, moved parse expression under parse

File: core/src/main/java/org/opensearch/sql/expression/parse/GrokExpression.java (hunk 4 above)
Comment (joshuali925) [reply core/src/main/java/org/opensearch/sql/expression/parse/GrokExpression.java:71]: it's to extract new derived fields to define in currEnv in analyzer

File: docs/user/ppl/cmd/patterns.rst (hunk 5 above)
Comment (joshuali925) [reply docs/user/ppl/cmd/patterns.rst]: fixed

File: docs/user/ppl/cmd/patterns.rst (hunk 6 above)
Comment (joshuali925) [reply docs/user/ppl/cmd/patterns.rst]: removed

File: common/src/main/java/org/opensearch/sql/common/grok/Discovery.java [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+    // Boolean done = false;
+    String texte = text;
+    GrokCompiler compiler = GrokCompiler.newInstance();
+    compiler.register(grokPatterns);
+
+    // Compile the pattern
+    for (Entry<String, String> stringStringEntry : grokPatterns.entrySet()) {
+      @SuppressWarnings("rawtypes")
+      Entry pairs = (Entry) stringStringEntry;
+      String key = pairs.getKey().toString();
```
Comment (MaxKsyunz) [inline common/src/main/java/org/opensearch/sql/common/grok/Discovery.java:106]: Why is this necessary?

Comment (joshuali925) [reply common/src/main/java/org/opensearch/sql/common/grok/Discovery.java:106]: [java-grok](https://github.com/thekrakken/java-grok) is an independent library that's used for the PPL's `grok` command, but in the latest release `Grok` is not serializable (and SQL grok expression cannot be pushed down). After discussing with peng I'm tentatively moving the source code into sql.common module.

And for this line, yes it's not required. Historically it was an iterator `Map.Entry pairs = (Map.Entry) it.next();`, and then was refactored to for each loop but that line was not removed. Actually the Discovery part is not used, will remove.

File: core/src/main/java/org/opensearch/sql/planner/physical/ProjectOperator.java [hunk 8]
```diff
@@ -65,10 +65,14 @@ public ExprValue next() {
...
     // TODO needs a better implementation, see https://github.com/opensearch-project/sql/issues/458
     for (NamedExpression expr : namedParseExpressions) {
-      ExprValue value = inputValue.bindingTuples()
-          .resolve(((ParseExpression) expr.getDelegated()).getExpression());
-      if (value.isMissing()) {
-        // value will be missing after stats command, read from inputValue if it exists
+      if (projectList.stream()
+          .noneMatch(field -> field.getNameOrAlias().equals(expr.getNameOrAlias()))) {
+        continue;
+      }
```
Comment (penghuo) [inline core/src/main/java/org/opensearch/sql/planner/physical/ProjectOperator.java:71]: Which case covered, could you explain more?

File: core/src/main/java/org/opensearch/sql/expression/parse/GrokExpression.java (hunk 2 above)
Comment (penghuo) [reply core/src/main/java/org/opensearch/sql/expression/parse/GrokExpression.java]: we should replace rawString as `***`.

**Review by penghuo — COMMENTED** [review]
files in common/test/resources/nasa is required? seems like split file

File: core/src/main/java/org/opensearch/sql/planner/physical/ProjectOperator.java (hunk 8 above)
Comment (joshuali925) [reply core/src/main/java/org/opensearch/sql/planner/physical/ProjectOperator.java:71]: 1. add all expressions in project list that are not parse expressions
2. add all parse expressions that are also in project list
     1. if parse expression is not pushed down as aggs, run parse logic in memory
     1. if parse expression is pushed down as aggs (stats), read it from inputValue

**Comment by joshuali925** [comment]
> files in common/test/resources/nasa is required? seems like split file

yes it's split file, used here
https://github.com/opensearch-project/sql/blob/f331e09c44ed895a0823c44330c9cb621a6c1f70/common/src/test/java/org/opensearch/sql/common/grok/ApacheTest.java#L52-L67

File: core/src/main/java/org/opensearch/sql/expression/parse/GrokExpression.java (hunk 2 above)
Comment (joshuali925) [reply core/src/main/java/org/opensearch/sql/expression/parse/GrokExpression.java]: hard coded it to `***`

File: core/src/main/java/org/opensearch/sql/expression/parse/ParseExpression.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+  public ExprType type() {
+    return ExprCoreType.STRING;
+  }
+
+  @Override
+  public <T, C> T accept(ExpressionNodeVisitor<T, C> visitor, C context) {
+    return visitor.visitParse(this, context);
+  }
+
+  abstract ExprValue parseValue(ExprValue value) throws ExpressionEvaluationException;
```
Comment (dai-chen) [inline core/src/main/java/org/opensearch/sql/expression/parse/ParseExpression.java:78]: np: is `protected` sufficient?

File: docs/user/ppl/cmd/patterns.rst [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+.. rubric:: Table of contents
+
+.. contents::
+   :local:
+   :depth: 2
+
+
+Description
+============
+| The ``patterns`` command extracts log patterns from a text field and appends the results to the search result.
```
[the remark this answers is not in the document]
Comment (joshuali925) [reply docs/user/ppl/cmd/patterns.rst]: added more description
> Grouping logs by their patterns makes it easier to aggregate stats from large volumes of log data for analysis and troubleshooting.

for details you can check this issue https://github.com/opensearch-project/observability/issues/893, if you have any suggestion feel free to let me know

File: docs/user/ppl/cmd/grok.rst [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    +------------------+
+
+Example 3: Using grok to parse logs
+===================================
+
+The example shows how to use grok to parse raw logs.
+
+PPL query::
+
+    os> source=accounts | eval raw='145.128.75.121 - - [29/Aug/2022:13:26:44 -0700] "GET /deliverables HTTP/2.0" 501 2721' | grok raw '%{COMMONAPACHELOG}' | head 1 | fields COMMONAPACHELOG, timestamp, response, bytes ;
```
[the remark this answers is not in the document]
Comment (joshuali925) [reply docs/user/ppl/cmd/grok.rst:73]: `head 1` because the source is set by eval, without it all lines will contain the same information. Actually might be better to use a new index instead of `accounts`, will check how to do that

File: core/src/main/java/org/opensearch/sql/expression/parse/ParseExpression.java (hunk 9 above)
Comment (joshuali925) [reply core/src/main/java/org/opensearch/sql/expression/parse/ParseExpression.java:78]: could you explain a bit more? i think it's default not protected, and should only be accessed by sub classes in the same package

File: docs/user/ppl/cmd/grok.rst (hunk 11 above)
[the remark this answers is not in the document]
Comment (joshuali925) [reply docs/user/ppl/cmd/grok.rst]: updated

File: core/src/main/java/org/opensearch/sql/expression/parse/ParseExpression.java (hunk 9 above)
Comment (dai-chen) [reply core/src/main/java/org/opensearch/sql/expression/parse/ParseExpression.java:78]: Okay, got it!

**Review by dai-chen — APPROVED** [review]
Thanks for the changes!

File: ppl/src/main/antlr/OpenSearchPPLParser.g4 [hunk 12]
```diff
@@ -94,10 +94,27 @@ rareCommand
...
     ;
-    
+
+patternsCommand
+    : PATTERNS (patternsParameter)* (source_field=expression)
+    ;
+
+patternsParameter
+    : (NEW_FIELD EQUAL new_field=stringLiteral)
+    | (PATTERN EQUAL pattern=stringLiteral)
```
Comment (vamsi-amazon) [inline ppl/src/main/antlr/OpenSearchPPLParser.g4:111]: What if the pattern itself contains single quotes and double quotes.

**Review by penghuo — APPROVED** [review]
